### PR TITLE
[RDY] Initial implementation of constants manager and moving NO_VITAMINS to game_constant from world option

### PR DESCRIPTION
**Pull request by Firestorm01X2**
Introducing initial implementation of constants manager.
What is it? It is simple way of storing game constants in JSONs and getting their values.

Using is pretty simple;
1) Explose your constant in JSON ( recommended: <code> game_constants.json </code>):
<code> 
{
"type" : "game_constant",
"id" : "MY_CONSTANT",
"description" : "Example of constant.",
stype": "float",
"value" : 1.5
 },</code>

2) Get it from your game code:
<code>
#include "constants_manager.h"
...

float myConstant = get_constant< **float** >(  "**MY_CONSTANT**" )
</code>

That is it. No more hidden world options. No more relations to save files.

Features:
*  Simple and quick way of exposing game constants to JSON 
* Simple to use
* Supports <code>string</code>, <code>bool</code>, <code>float</code> and <code>int</code> constants
* Supports modding
* Some constants from game_constants.h can go there 
* Very useful for mods like NO_VITAMINS or NO_NPC_FOOD. No more pollution of world option with something that is not supposed to be option at all and no more storing that in save directory
* Useful in debugging
* Constants support multiple loadings so in can be easily exposed to mods

Mod example:
NO_VITAMINS that is changed in this PR.

### CODE REVIEW

File: data/json/game_constants.json [hunk 1]
```diff
@@ -0,0 +1,30 @@
+[
+    {
+        "type" : "game_constant",
+		"id" : "STRING_CONSTANT_EXAMPLE",
```
Comment (Night-Pryanik) [inline data/json/game_constants.json]: Indentation here and below.

File: data/json/game_constants.json [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        "description" : "Example of string constant.",
+        "stype": "string",
```
Comment (Night-Pryanik) [inline data/json/game_constants.json]: Colon should be at one whitespace from the type field.

File: src/constants_manager.cpp [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+{
+    static game_constant null_constant;
+    auto iter = game_constants_all.find(id);
+    return iter != game_constants_all.end() ? iter->second : null_constant;
+}
+
+template<>
+std::string constants_manager::game_constant::value_as<std::string>() const
+{
+    if ( stype_ != "string" ) {
```
Comment (Night-Pryanik) [inline src/constants_manager.cpp]: Astyle all ifs here and below like `if( xxx )`.

**Comment by Firestorm01X2** [comment]
Styling done. Also Visual studio successfully built something that g++ cannot. That is fixed also.

**Comment by Night-Pryanik** [comment]
Styling isn't completed unfortunately. Please run astyle on your changes.

**Comment by Firestorm01X2** [comment]
Now it is done and Astyled.

**Comment by ZhilkinSerg** [comment]
Can you provide an example mod?

**Comment by Firestorm01X2** [comment]
Idea itself is simple and usual. You write constant with same id in some json  in directory of your mod and it ovewrites original constant. 
Of corse constant value should be used in game code so it make sense.

**Comment by Firestorm01X2** [comment]
>Can you provide an example mod?
Mod example attached.

**Comment by ZhilkinSerg** [comment]
> Mod example attached.

No, I meant real usage example.

**Comment by Firestorm01X2** [comment]
>No, I meant real usage example.

For example we have in item.cpp:
<code>
// Dividing dispersion by 3 temporarally as a gross adjustment,
    // will bake that adjustment into individual gun definitions in the future.
    // Absolute minimum gun dispersion is 45.
    dispersion_sum = std::max( dispersion_sum / 3, 45 );
</code>
I can replace it with:
 <code>dispersion_sum = std::max( dispersion_sum / **get_constant< float >( "RANGED_DIV" )**, 45 );</code>

After that I will write in game_constants:
<code> { "type" : "game_constant", "id" : "RANGED_DIV", "description" : "Ranged divider.", stype": "float", "value" : 3 }</code>

After that you can write mod will changes RANGED_DIV.

**Comment by Firestorm01X2** [comment]
Here. For example I've remove NO_VITAMINS from WORLD OPTIONs and put it to game constants. Also I've changed NO_VITAMINS mod to override game_constants. That is good example.

**Comment by kevingranade** [comment]
I'm at a loss for how this improves anything.  What can this do that an option can't?

Why is it not being saved a good thing? The existing options get saved along with the world so that a player can select different options for different worlds, this forces them to use the same options for all worlds, which is almost never what I'd expect to be the desired outcome, especially if a player decides to change a setting mid-game, or when they start a new game while continuing to play in an old game world.

File: data/json/game_constants.json [hunk 4]
```diff
@@ -0,0 +1,37 @@
+[
+  {
+    "type": "game_constant",
```
Comment (kevingranade) [inline data/json/game_constants.json]: Remove the non-functional examples, they're just clutter.

File: src/constants_manager.h [hunk 5]
```diff
@@ -0,0 +1,58 @@
+#pragma once
+#ifndef CONSTANTS_MANAGER_H
+#define CONSTANTS_MANAGER_H
+
+#include "json.h"
```
Comment (BevapDin) [inline src/constants_manager.h]: You don't need to include "json.h" - a forward declaration of `JsonObject` is enough.

You also don't need to include "set", but you should include "map" and "string" as the header uses those.

File: src/constants_manager.cpp [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+
+constants_manager &get_constants_manager()
+{
+    static constants_manager single_instance;
+    return single_instance;
+}
+
+
+constants_manager::game_constant &constants_manager::get_constant( const std::string &id )
```
Comment (BevapDin) [inline src/constants_manager.cpp]: This function should be `const`. And returning a reference to a non-const `game_constant` looks strange. Does it need to be non-const?

File: src/constants_manager.cpp [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    if( stype_ == "float" ) {
+        float_value_ = jo.get_float( "value" );
+    } else if( stype_ == "int" ) {
+        int_value_ = jo.get_int( "value" );
+    } else if( stype_ == "bool" ) {
+        bool_value_ = jo.get_bool( "value" );
+    } else if( stype_ == "string" ) {
+        string_value_ = jo.get_string( "value" );
+    } else {
+        debugmsg( "Unknown stype for %s.", id_.c_str() );
```
Comment (BevapDin) [inline src/constants_manager.cpp]: Call `jo.error( "your message", "stype" )` to get a proper error message.

File: src/constants_manager.h [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+                template<typename T>
+                T value_as() const;
+                game_constant( const std::string &id = std::string() ) : id_( id ) {}
+
+                /** Load constant definition from JSON */
+                void load( JsonObject &jo );
+
+            private:
+                const std::string id_;
+                std::string description_;
```
Comment (BevapDin) [inline src/constants_manager.h]: What does the description? It seems unused. Is it really needed in the code?

File: src/constants_manager.h [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+                std::string stype_;
```
Comment (BevapDin) [inline src/constants_manager.h]: I know option manger uses a string for this as well, but it should be an enumeration.

**Comment by Firestorm01X2** [comment]
>I'm at a loss for how this improves anything. What can this do that an option can't?

Main idea that is is for constants. So no changing  in mid game. No changing via code.

Hidden options basically same thing as cosntant but it is stored in save. But for what? Player can't (must not) chage it in game. Also there is no reason change it in game code.

File: src/constants_manager.h (hunk 8 above)
Comment (Firestorm01X2) [reply src/constants_manager.h]: It may be useful for observation and analysis. And somewhat force to explain what is it and why do we need it.

Comment (Firestorm01X2) [reply src/constants_manager.h]: Comments can do it too. But comments can't be observed in game . Also modder can write info here about why do they overwrite that constant

**Comment by Firestorm01X2** [comment]
Done.

**Comment by Firestorm01X2** [comment]
Now it is done.

**Comment by kevingranade** [comment]
> Main idea that it is for constants. So no changing in mid game. No changing via code. Just use values.

So exactly the same as hidden options. If there is no additional functionality, there is no reason to have this code.

**Comment by Firestorm01X2** [comment]
That  is just great.

**Pull request closed without merging**